### PR TITLE
Extend model serialization test to all backbones

### PR DESCRIPTION
**Pull request by soma2000-lang**
Address #1387
@jbischof

### CODE REVIEW

File: keras_cv/models/models_test.py [hunk 1]
```diff
@@ -156,6 +158,18 @@ def _test_model_can_be_used_as_backbone(self, app, last_dim, args):
         model = keras.Model(inputs=inputs, outputs=[backbone_output])
         model.compile()
 
+    def _test_model_serialization(self, app, _, args, save_format, filename):
```
Comment (jbischof) [inline keras_cv/models/models_test.py:161]: This code should already be available in #1367

**Review by jbischof — APPROVED** [review]
Looks great thank you!

File: keras_cv/models/models_test.py (hunk 1 above)
Comment (ianstenbit) [reply keras_cv/models/models_test.py:161]: +1 -- @soma2000-lang could you please merge from master@head and then re-push? otherwise lgtm.

**Review by ianstenbit — APPROVED** [review]
Thank you!

Left one comment about Git history, but once that's fixed we can /gcbrun and merge!

**Comment by soma2000-lang** [comment]
image


@jbischof @ianstenbit I am trying to do that,but to no help.Please help me out.

Regards

**Comment by jbischof** [comment]
Did you `git fetch --all`?

**Comment by ianstenbit** [comment]
> Did you `git fetch --all`?

You probably need to add an upstream branch and merge from there

```
git remote add upstream https://github.com/keras-team/keras-cv.git
git fetch upstream
git merge upstream/master
```

**Comment by soma2000-lang** [comment]
@jbischof @ianstenbit Thanks Done!

**Comment by ianstenbit** [comment]
@soma2000-lang just a quick update:

It looks like this is causing our CI to timeout (probably because of the extra time necessary to serialize/deserialize all of these models).

Also, we may wait to include the `"keras_v3"` format depending on the outcome of #1390 (as the v3 format may depend on Keras 2.12, which we don't want to introduce a dependency on yet)

**Comment by jbischof** [comment]
Yes @soma2000-lang I would recommend changing the tests to:
```python
    @parameterized.parameters(*MODEL_LIST)
    def test_model_serialization_tf(self, app, last_dim, args):
        super()._test_model_serialization(
            app, last_dim, args, save_format="tf", filename="model"
        )

    @parameterized.parameters(*MODEL_LIST)
    def test_model_serialization_keras_format(self, app, last_dim, args):
        if version.parse(tf.__version__) >= version.parse("2.12.0-dev0"):
            super()._test_model_serialization(
                app, last_dim, args, save_format="keras_v3", filename="model.keras"
            )
```

**Comment by soma2000-lang** [comment]
@jbischof @ianstenbit Done!

**Comment by ianstenbit** [comment]
@soma2000-lang please take a look at the code format issue. Thank you!

**Comment by soma2000-lang** [comment]
@jbischof  I have reformatted the code using "sh shell/format.sh",still cannot contemplate why the tests code format test are failing

**Comment by ianstenbit** [comment]
@soma2000-lang most likely you need to merge past #1402

**Comment by soma2000-lang** [comment]
@ianstenbit @jbischof Should I just write #noqa in the line exceeding 80 chars?

**Comment by ianstenbit** [comment]
> @ianstenbit @jbischof @LukeWood Should I just write #noqa in the lines exceeding 80 chars?

It looks like the issue is:
```
ERROR: /home/runner/work/keras-cv/keras-cv/keras_cv/models/resnet_v1_test.py Imports are incorrectly sorted and/or formatted.
```

I don't see any line length issues but maybe I'm missing something.

File: keras_cv/models/resnet_v1_test.py [hunk 2]
```diff
@@ -14,7 +14,9 @@
 
 import tensorflow as tf
 from absl.testing import parameterized
-from packaging import version
+
+from packaging.version import Version
+
```
Comment (james77777778) [inline keras_cv/models/resnet_v1_test.py:19]: There are 2 empty lines around `from packaging.version import Version`.

`./shell/format.sh` can fix this automatically

**Comment by soma2000-lang** [comment]
@james77777778 I had already given the command.

**Comment by james77777778** [comment]
Hi @soma2000-lang 
I find some bugs in `keras_cv/models/resnet_v1_test.py` of this PR

1. There are some empty lines around `from packaging.version import Version`[https://github.com/soma2000-lang/keras-cv/blob/model/keras_cv/models/resnet_v1_test.py#L17-L19](https://github.com/soma2000-lang/keras-cv/blob/model/keras_cv/models/resnet_v1_test.py#L17-L19)
2. `test_model_serialization_tf` and `test_model_serialization_keras_format` are duplicated [https://github.com/soma2000-lang/keras-cv/blob/model/keras_cv/models/resnet_v1_test.py#L56-L84](https://github.com/soma2000-lang/keras-cv/blob/model/keras_cv/models/resnet_v1_test.py#L56-L84)

`./shell/format.sh` should work and remove empty lines automatically. `keras_cv/models/resnet_v1_test.py` was reformated after I execute the script on your code.

```python
import tensorflow as tf
from absl.testing import parameterized
from packaging.version import Version

from keras_cv.models import resnet_v1

from .models_test import ModelsTest

MODEL_LIST = [
    (resnet_v1.ResNet18, 512, {}),
]

"""
Below are other configurations that we omit from our CI but that can/should
be tested manually when making changes to this model.
(resnet_v1.ResNet34, 512, {}),
(resnet_v1.ResNet50, 2048, {}),
(resnet_v1.ResNet101, 2048, {}),
(resnet_v1.ResNet152, 2048, {}),
"""


class ResNetV1Test(ModelsTest, tf.test.TestCase, parameterized.TestCase):
    @parameterized.parameters(*MODEL_LIST)
    def test_application_base(self, app, _, args):
        super()._test_application_base(app, _, args)

    @parameterized.parameters(*MODEL_LIST)
    def test_application_with_rescaling(self, app, last_dim, args):
        super()._test_application_with_rescaling(app, last_dim, args)

    @parameterized.parameters(*MODEL_LIST)
    def test_application_pooling(self, app, last_dim, args):
        super()._test_application_pooling(app, last_dim, args)

    @parameterized.parameters(*MODEL_LIST)
    def test_application_variable_input_channels(self, app, last_dim, args):
        super()._test_application_variable_input_channels(app, last_dim, args)

    @parameterized.parameters(*MODEL_LIST)
    def test_model_serialization_tf(self, app, last_dim, args):
        super()._test_model_serialization(
            app, last_dim, args, save_format="tf", filename="model"
        )

    @parameterized.parameters(*MODEL_LIST)
    def test_model_serialization_keras_format(self, app, last_dim, args):
        if Version.parse(tf.__version__) >= Version.parse("2.12.0-dev0"):
            super()._test_model_serialization(
                app,
                last_dim,
                args,
                save_format="keras_v3",
                filename="model.keras",
            )


if __name__ == "__main__":
    tf.test.main()

```

Hope this helps!

**Comment by james77777778** [comment]
Hi @soma2000-lang ,
I did some survey based on your PR.

From previous implementation
https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/resnet_v1_test.py#L17
the import is `from packaging import version`

But in the PR `from packaging.version import Version` is used. 

I think `Version.parse(...)` in tests should raise `AttributeError`?
```bash
>>> from packaging import version
>>> from packaging.version import Version
>>> version.parse("2.12.0-dev0")
<Version('2.12.0.dev0')>
>>> Version.parse("2.12.0-dev0")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Version' has no attribute 'parse'
```

Hi @ianstenbit 
I'm curious why CI tests did not catch these errors?

Ref:
[https://packaging.pypa.io/en/stable/version.html](https://packaging.pypa.io/en/stable/version.html)

**Comment by soma2000-lang** [comment]
@jbischof @ianstenbit finally I have fixed the issue

**Comment by ianstenbit** [comment]
@james77777778 -- the base CI tests don't run model tests, only the accelerator tests run these (triggered by /gcbrun), which is likely why we weren't seeing CI failures for that issue you pointed out.

Thank you both for the discussion and contribution!

**Comment by ianstenbit** [comment]
@soma2000-lang it looks like the cost of these tests is resulting in timeouts for our GCB tests.

One thing we could try is to update the shape of the inputs in the base class. Currently, the base model test uses `(16, 224, 224, 3)` as the shape of the inputs for the serialization test, but we could probably use `(1, 224, 224, 3)` instead.

I'm not sure how much better this would be but it's something to try.

**Comment by soma2000-lang** [comment]
@ianstenbit sure will do the change!

**Comment by ianstenbit** [comment]
@soma2000-lang please take a look at the CI failures.

Then we can re-start the accelerator tests to see if they're a bit speedier now.

**Comment by soma2000-lang** [comment]
@ianstenbit Made the changes but not sure why the CI tests are not running

**Comment by jbischof** [comment]
@ianstenbit @soma2000-lang I'm sorry I didn't realize this PR is still in flight. The purpose of this task was to verify serialization when the backbones were migrated to functional subclass, but that is moot at this point.

I wonder if we should abandon at this point in favor of the new testing style in #1458 when we update all the backbones. Then we likely should hide the test behind a pytest `--run_large` or `--run_extra_large` flag given the timeout issues.

**Comment by ianstenbit** [comment]
> @ianstenbit @soma2000-lang I'm sorry I didn't realize this PR is still in flight. The purpose of this task was to verify serialization when the backbones were migrated to functional subclass, but that is moot at this point.
> 
> I wonder if we should abandon at this point in favor of the new testing style in #1508 when we update all the backbones. Then we likely should hide the test behind a pytest `--run_large` or `--run_extra_large` flag given the timeout issues.

I agree that at this point we should probably just wait for the `Backbone` PRs for this change, and have them made 1x1.

@soma2000-lang my apologies for this -- we really appreciate your effort on this, and your many other contributions to KerasCV!

**Pull request closed without merging**